python_modules/libraries/dagster-fivetran/dagster_fivetran/v2/translator.py

```diff
@@ -0,0 +1,81 @@
+from enum import Enum
+from typing import Any, Mapping, Sequence
+
+from dagster import _check as check
+from dagster._core.definitions.asset_key import AssetKey
+from dagster._core.definitions.asset_spec import AssetSpec
+from dagster._record import record
+
+
+class FivetranContentType(Enum):
+    """Enum representing each object in Fivetran's ontology."""
+
+    CONNECTOR = "connector"
+    DESTINATION = "destination"
+
+
+@record
+class FivetranContentData:
+    """A record representing a piece of content in a Fivetran workspace.
+    Includes the object's type and data as returned from the API.
+    """
+
+    content_type: FivetranContentType
+    properties: Mapping[str, Any]
+
+    def to_cached_data(self) -> Mapping[str, Any]:
+        return {"content_type": self.content_type.value, "properties": self.properties}
+
+    @classmethod
+    def from_cached_data(cls, data: Mapping[Any, Any]) -> "FivetranContentData":
+        return cls(
+            content_type=FivetranContentType(data["content_type"]),
+            properties=data["properties"],
+        )
+
+
+@record
+class FivetranWorkspaceData:
+    """A record representing all content in a Fivetran workspace.
+    Provided as context for the translator so that it can resolve dependencies between content.
+    """
+
+    connectors_by_id: Mapping[str, FivetranContentData]
+    destinations_by_id: Mapping[str, FivetranContentData]
+
+    @classmethod
+    def from_content_data(
+        cls, content_data: Sequence[FivetranContentData]
+    ) -> "FivetranWorkspaceData":
+        raise NotImplementedError()
+
+
+class DagsterFivetranTranslator:
+    """Translator class which converts raw response data from the Fivetran API into AssetSpecs.
+    Subclass this class to implement custom logic for each type of Fivetran content.
+    """
+
+    def __init__(self, context: FivetranWorkspaceData):
+        self._context = context
+
+    @property
+    def workspace_data(self) -> FivetranWorkspaceData:
+        return self._context
+
+    def get_asset_key(self, data: FivetranContentData) -> AssetKey:
+        if data.content_type == FivetranContentType.CONNECTOR:
+            return self.get_connector_asset_key(data)
+        else:
+            check.assert_never(data.content_type)
+
+    def get_asset_spec(self, data: FivetranContentData) -> AssetSpec:
+        if data.content_type == FivetranContentType.CONNECTOR:
+            return self.get_connector_spec(data)
+        else:
+            check.assert_never(data.content_type)
+
+    def get_connector_asset_key(self, data: FivetranContentData) -> AssetKey:
+        raise NotImplementedError()
+
+    def get_connector_spec(self, data: FivetranContentData) -> AssetSpec:
```

Comment:
I'm assuming this might be inherited from existing patterns, but I do find this kinda awkward. Creates a weird footgun where you can implement get_connector_spec in a way that does _not_ utilize get_connector_asset_key, right? 

I wonder if we should _just_ have `get_connector_spec` and then force people to use `get_connector_spec(...).asset_key`.

To make this more efficient, we could cache the get_connector_spec result per datum.

Reply:
> I'm assuming this might be inherited from existing patterns, but I do find this kinda awkward.

Yes exactly. A similar discussion came up [here](https://github.com/dagster-io/dagster/pull/25432#discussion_r1813887047).

I'm in favor of this change - I agree that this pattern is error prone. 

It was kinda easier for a user to override the `get_asset_key` function, to add a prefix for instance. We could recommend doing something like the following instead, which looks a bit awkward but is less error prone. 

```python
class CustomDagsterTranslator(DagsterTranslator):
    def get_asset_spec(self, props) -> AssetSpec:
        asset_spec = super().get_asset_spec(props)
        return asset_spec._replace(asset_key=asset_spec.asset_key.with_prefix("my_prefix"))
```

Reply:
I think we have better fxns for updating params on asset specs now too: `replace_attributes` top level fxn. But yea I think this pattern is good. Having an example people can copy paste, even if it looks slightly awkward, is much preferable to introducing a footgun.

Reply:
Have we made the other integrations where this fxn exists public yet @maximearmstrong ? / would like to understand the status / experimentality there.

Reply:
I'm on board with this change - I think our initial push for `get_asset_key` was to avoid doing extra work to construct an entire asset spec when we would discard most of it & just retrieve the key, but Chris made a good point that any asset whose key we're generating we'll also be generating a spec for, either before or after. As long as we can cache the spec generation, we don't end up doing extra work in that case.

Reply:
The `get_asset_key` method is also not marked as experimental for the BI integrations, so that implies a deprecation cycle, but the risk seems pretty low here.

Reply:
Yea if we're going to do it, we should probably do so soon. Want to post a public discussion about it to make sure we get buy in from @schrockn @benpankow ; do we have anyone customizing BI translators to our knowledge yet?

Reply:
To my knowledge, no one is customizing the Tableau and Power BI translators yet. @benpankow Do you know if anyone is customizing the Sigma and Looker ones?

I can post the public discussion with code snippets to get consensus on this. 

For this PR, I will remove the `get_asset_key` method - we can easily add it depending on the outcome of the discussion.

Reply:
yep that sounds good to me!